appsec-kit-v7/src/main/java/com/vaadin/appsec/v7/service/NotificationInitializer.java

```diff
@@ -37,91 +36,49 @@ public class NotificationInitializer {
     private static final Logger LOGGER = LoggerFactory
             .getLogger(NotificationInitializer.class);
 
-    private static final String NOTIFIED_UIS_SESSION_PARAM = "vaadin-appsec-kit-notified-uis";
-    /**
-     * New UI instance check interval (in ms). Determines how often the session
-     * is checked for new UI instances which might need a notification shown.
-     */
-    private static final int NOTIFICATION_CHECK_INTERVAL = 5000;
-
     /**
      * Notification timeout (in ms). Set to 24 hours to essentially make it
      * persistent until either the appsec link is clicked or the notification is
      * dismissed.
      */
     private static final int NOTIFICATION_DELAY = 24 * 60 * 60 * 1000;
 
-    private static final Map<String, ScheduledFuture<?>> scheduledChecks = new ConcurrentHashMap<>();
+    private final Map<VaadinSession, Registration> scanEventRegistrations = new ConcurrentHashMap<>();
 
-    /**
-     * Initializes the notification service
-     *
-     * @param vaadinService
-     *            current VaadinService instance
-     */
-    static void serviceInit(VaadinService vaadinService) {
-        if (isDebugMode(vaadinService)) {
-            vaadinService.addSessionInitListener(e -> {
-                e.getSession().addUIProvider(new AppSecUIProvider());
-                startNotificationTask(e.getSession());
-            });
-            LOGGER.info("NotificationInitListener initialized.");
+    void serviceInit(VaadinService service) {
+        if (isDebugMode(service)) {
+            service.addSessionInitListener(this::subscribeSessionToScanEvents);
+            service.addSessionDestroyListener(this::removeSessionRegistration);
+            LOGGER.info("Subscribed to AppSec Kit scan events");
         }
     }
 
-    private static void startNotificationTask(final VaadinSession session) {
-        final String sessionId = session.getSession().getId();
-        final ScheduledFuture<?> scheduledNotificationCheck = AppSecService
-                .getInstance().getConfiguration().getTaskExecutor()
-                .scheduleAtFixedRate(() -> {
-                    if (isSessionOpen(session)) {
-                        session.access(() -> {
```

Comment:
During testing this branch without this `session.access(...)` function, which provides exclusive access to this session from outside a request handling thread, the notification popup's behaviour was nondeterministic. It didn't popup during the application initialization after the first scan and there were errors when I opened and closed more sessions after each other.
With using this function the notification popup appeared when expected.

Reply:
True, my intention was to replace that with `UI::access` but then I forgot to do so. Thanks!